src/Documentation/Event-Sourcing/LogConsistencyProviders.md

```diff
@@ -0,0 +1,54 @@
+---
+layout: page
+title: Included Log-Consistency Providers
+---
+
+# Built-In Log-Consistency Providers
+
+The `Microsoft.Orleans.EventSourcing` package includes several log-consistency providers that cover basic scenarios suitable to get started, and allow some extensibility.
+
+
+### Orleans.EventSourcing.**StateStorage**.LogConsistencyProvider
+
+This provider stores *grain state snapshots*, using a standard storage provider that can be configured independently. 
+
+The data that is kept in storage is an object that contains both the grain state (as specified by the first type parameter to `JournaledGrain`) and some meta-data (the version number, amd a special tag that is used to avoid duplication of events when storage accesses fail).
+
+Since the entire grain state is read/written every time we access storage, this provider is not suitable for objects whose grain state is very large. 
+
+This provider does not support `RetrieveConfirmedEvents`: it cannot retrieve the events from storage because the events are not persisted.
```

Comment:
> Typically replaying of the events does not happen at normal operation time
I was originally assuming that as well, but I now think that it really depends on what kind of system you want to build. I don't think it's always true. More importantly, I don't think enforcing this makes anybody's life easier.

> Adding this method by itself can lead to hard to solve data consistency issues.
I dont' see why. The confirmed events never change.

> You won't go grain by grain as they get normally activated (you might though), but instead have a startup process that goes through the entire log and re-process

This seems like an advanced scenario. There is currently no way to write such a provider in a general way (since it requires knowing all the grains and how the grain state was changed).

I think we should support the simplest solution first: for each grain, when grain is loaded, restore from most recent checkpoint found (if possible) or otherwise (no checkpoint found or checkpoint is no longer valid due to some changes) replay entire log. I don't really see a problem with this.

The only thing needed to support this is some way to know if a checkpoint is invalid, e.g. because the grain state has changed (e.g. different fields), or the user has changed the transition function (e.g. fixed a bug). I am not sure there is an easy way to detect this automatically - clearly, a deserialization failure would indicate that the checkpoint is useless. Perhaps this is enough, as users can plug in their own deserialization code and use an explicit versioning mechanism.